### PR TITLE
feat: import/export of config and secrets

### LINKED ISSUE #1948: Add a way to dump & restore secrets
Could be a script or a proper cmd
if a command, will probably need to take in provider key so we don't end up dumping and restoring everything into 1password instead of just migrating 1password secrets

### PR DESCRIPTION
**Pull request by matt2e**
closes #1948

**Usage:**

**Exporting**
Prints json of all secrets to stdout:
`ftl secret export`

Exports all inline secrets to a json file
`ftl secret export --inline > exported.json`

**Importing**
Imports json from a file:
`ftl secret import --inline exported.json`

**Migrating**
Migrates inline secrets to 1Password:
`ftl secret export --inline | ftl secret import --op --opvault <vauld id>`

**JSON Format**
```
{
    "<module>.<name>": <secret>,
    ...
}
```

### CODE REVIEW

File: cmd/ftl/cmd_secret.go [hunk 1]
```diff
@@ -19,10 +19,12 @@ import (
 )
 
 type secretCmd struct {
-	List  secretListCmd  `cmd:"" help:"List secrets."`
-	Get   secretGetCmd   `cmd:"" help:"Get a secret."`
-	Set   secretSetCmd   `cmd:"" help:"Set a secret."`
-	Unset secretUnsetCmd `cmd:"" help:"Unset a secret."`
+	List   secretListCmd   `cmd:"" help:"List secrets."`
```
Comment (alecthomas) [inline cmd/ftl/cmd_secret.go:22]: Let's replicate this for config as well - the two sets of commands should be almost identical.

File: cmd/ftl/cmd_secret.go [hunk 2]
```diff
@@ -169,3 +171,91 @@ func (s *secretUnsetCmd) Run(ctx context.Context, scmd *secretCmd, adminClient a
 	}
 	return nil
 }
+
+type secretImportCmd struct {
+	Input *string `arg:"" placeholder:"JSON" help:"JSON to import as secrets (read from stdin if omitted). Format: [{\"ref\":\"<module>.<name>\",\"value\": <secret>}, ...]" optional:""`
```
Comment (alecthomas) [inline cmd/ftl/cmd_secret.go:176]:
```suggestion
	Input *os.File `arg:"" placeholder:"JSON" help:"JSON to import as secrets (read from stdin if omitted). Format: [{\"ref\":\"<module>.<name>\",\"value\": <secret>}, ...]" optional:"" default:"-"`
```

^^ this should do what the below manual code does

File: cmd/ftl/cmd_secret.go [hunk 3]
```diff
@@ -169,3 +171,91 @@ func (s *secretUnsetCmd) Run(ctx context.Context, scmd *secretCmd, adminClient a
...
+	var err error
+	if s.Input != nil {
+		input = []byte(*s.Input)
+	} else {
+		input, err = io.ReadAll(os.Stdin)
+		if err != nil {
+			return fmt.Errorf("failed to read config from stdin: %w", err)
+		}
+	}
+	var entries map[string]any
```
Comment (alecthomas) [inline cmd/ftl/cmd_secret.go:196]: This should just be a shared type that is used for both marshalling and unmarshalling, probably just `map[cf.Ref]string` (and implement `json.Marshaler`/`Unmarshaler` on `cf.Ref` if it isn't already).

Also the docs suggest it is an array, but this is a map?

Comment (alecthomas) [reply cmd/ftl/cmd_secret.go:196]: BTW generally, in Go, you should almost never be JSON marshalling/unmarshalling into `any`, if you know what the intended structure is.

File: cmd/ftl/cmd_secret.go [hunk 4]
```diff
@@ -169,3 +171,91 @@ func (s *secretUnsetCmd) Run(ctx context.Context, scmd *secretCmd, adminClient a
...
+		IncludeValues: optional.Some(true).Ptr(),
+	}
+	if provider, ok := scmd.provider().Get(); ok {
+		req.Provider = &provider
+	}
+	listResponse, err := adminClient.SecretsList(ctx, connect.NewRequest(req))
+	if err != nil {
+		return fmt.Errorf("could not retrieve secrets: %w", err)
+	}
+	entries := make(map[string]any, 0)
```
Comment (alecthomas) [inline cmd/ftl/cmd_secret.go:245]: Also not a JSON array as the docs suggest.

File: common/configuration/manager.go [hunk 5]
```diff
@@ -75,6 +76,10 @@ func New[R Role](ctx context.Context, router Router[R], providers []Provider[R])
 	return m, nil
 }
 
+func (m *Manager[R]) ProviderKeyForAccessor(accessor *url.URL) string {
```
Comment (alecthomas) [inline common/configuration/manager.go:79]: This can probably be a static function, but it is also fine.

File: go-runtime/compile/compile_integration_test.go [hunk 6]
```diff
@@ -15,9 +15,8 @@ func TestNonExportedDecls(t *testing.T) {
 		in.CopyModule("echo"),
 		in.Deploy("echo"),
 		in.CopyModule("notexportedverb"),
-		in.ExpectError(
-			in.ExecWithOutput("ftl", "deploy", "notexportedverb"),
-			"call first argument must be a function but is an unresolved reference to echo.Echo, does it need to be exported?"),
+		in.ExecWithExpectedError("call first argument must be a function but is an unresolved reference to echo.Echo, does it need to be exported?",
```
Comment (alecthomas) [inline go-runtime/compile/compile_integration_test.go:18]: Why the change? The previous function was generalisable to any Action.

File: cmd/ftl/cmd_secret.go (hunk 4 above)
Comment (matt2e) [reply cmd/ftl/cmd_secret.go:245]: Ah, yes the docs are out of date, will fix

File: cmd/ftl/cmd_secret.go (hunk 3 above)
Comment (matt2e) [reply cmd/ftl/cmd_secret.go:196]: This is a case where we would want any though right? Each value is arbitrary json, so I was trying to make the file format like this:
```json
{
   "echo.default_name": "hello world",
   "echo.info": {
          "key1": "value1",
          "key2": "value2"
   }
}
```

Comment (alecthomas) [reply cmd/ftl/cmd_secret.go:196]: Ah yes, true. However in that case you should use `json.RawMessage` instead of `any`.

File: go-runtime/compile/compile_integration_test.go (hunk 6 above)
Comment (matt2e) [reply go-runtime/compile/compile_integration_test.go:18]: I went in circles a bit on this:
- When I used `ExecWithOutput` I expected to be able to use the output. It didn't indicate that it was only capturing the output in case of an error
- I then tried to make `ExecWithOutput` work for both success success and error cases:
    - For success cases, call a func with the output (like the replacement `ExecWithOutput` action I added)
    - For error cases throw the error with the output (old behaviour)
- It then felt a bit messy for the error case so I tried cleaning it up.

I can go back to `ExecWithOutput` handling both cases. Thoughts?

Comment (alecthomas) [reply go-runtime/compile/compile_integration_test.go:18]: I think it's fine to introduce new functionality, but that doesn't seem like a reason to remove `in.ExpectError()`?

**Comment by alecthomas** [comment]
What's the behaviour when restoring, I assume it will just write the values in the dump, not remove any extra keys in the target?

**Comment by alecthomas** [comment]
FYI @AlexSzlavik

**Comment by matt2e** [comment]
> What's the behaviour when restoring, I assume it will just write the values in the dump, not remove any extra keys in the target?

That's right. Existing refs that aren't in the dump aren't touched. So it's less of a "restore" and of an "import"

File: go-runtime/compile/compile_integration_test.go (hunk 6 above)
Comment (matt2e) [reply go-runtime/compile/compile_integration_test.go]: Yeh I see what you mean. I've moved them back into one func called `ExecWithOutput` which works like before for the error case (used in conjunction with `ExpectError` and an empty capture func).

File: common/configuration/manager.go (hunk 5 above)
Comment (matt2e) [reply common/configuration/manager.go:79]: Already ran into a need for this to be static in another branch. Switching it over now.
